### ISSUE TITLE
"if condition is always false [redundant-expr]" false positive when variable set in a loop after the conditional

### ISSUE THREAD

**Issue by james-johnston-thumbtack** [python/mypy#14993]
**Bug Report**

The `redundant-expr` gives a false positive if you conditionally read a variable for another conditional inside a loop.

**To Reproduce**

```python
def mypytest() -> int | None:
    trigger_segment = None
    violations = None
    for segment in [1, 2]:
        if segment == 2:
            violations = trigger_segment if trigger_segment else segment
        trigger_segment = segment
    return violations
```

The rule will trigger with `error: If condition is always false  [redundant-expr]` on the `if trigger_segment` part of the ternary.

**Expected Behavior**

Mypy should not report anything because clearly the if condition is not always false: running `print(mypytest())` will print `1`.

**Actual Behavior**

The rule will trigger with `error: If condition is always false  [redundant-expr]` on the `if trigger_segment` part of the ternary.

**Your Environment**

My configuration:
```

[tool.mypy]
plugins = [
    "pydantic.mypy"
]
files = "src/**/*.py,tests/**/*.py"
strict = true
# These are additional checks not covered by strict:
disallow_any_unimported = true
enable_error_code = """\
    ignore-without-code,\
    truthy-bool,\
    truthy-iterable,\
    redundant-expr,\
    redundant-self,\
    unused-awaitable\
    """
warn_unreachable = true
```
```% mypy --version
mypy 1.1.1 (compiled: yes)
% python --version
Python 3.11.1
```

**Comment by Anexen** [comment]
I suppose that `trigger_segment` has inferred type `None`. You could try the following workaround:
```python 
trigger_segment: int | None = None
```

**Comment by hauntsaninja** [comment]
Fixed by #18433
